docs/architecture/0e-plans.md

```diff
@@ -0,0 +1,474 @@
+
+#### TF-PSA-Crypto header locations
+
+We distinguish between three categories of headers:
+
+* Public headers define public, stable APIs.
+* Exposed headers define exposed interfaces, as well as private interfaces used by Mbed TLS. They need to be present when building Mbed TLS.
+* Private headers are only needed when building TF-PSA-Crypto itself.
+
+The following table summarizes the characteristics of each category.
+
+| Category | Location | In include path? |
+| -------- | -------- | ---------------- |
+| Public   | `include` | yes |
+| Exposed  | `include` | yes except for Doxygen |
+| Private  | `drivers/*/include` | only for crypto |
+
+TODO: Task to set up the Doxygen build
+
+#### Hiding functions in an exposed header
+
+For TF-PSA-Crypto 0ε, all private functions defined in private headers will be guarded by `defined(MBEDTLS_DECLARE_PRIVATE_IDENTIFIERS)`. We define this macro when building TF-PSA-Crypto and Mbed TLS, but applications should not define it.
+
+This is similar to `MBEDTLS_ALLOW_PRIVATE_ACCESS` to “bless” access to structure fields.
+
+Prototype: https://github.com/Mbed-TLS/TF-PSA-Crypto/pull/132
+
+Thus, for the 0ε milestone, in each affected header, we just need to place `#ifdef` guards around the function declarations.
+
+Rationale:
+
+* This is a small amount of work.
+* This requires few changes to the code, and they are very localized, so it will not disrupt other work happening in parallel.
+
+TODO: this leaves Doxygen comments around
+
+#### Separating private interfaces from exposed interfaces
+
+If a private interface of TF-PSA-Crypto is declared in an exposed header, in the medium term, we should move it to a private header. Note that we can only do that if the interface is not called from Mbed TLS code.
+
+When none the functions declared in an exposed header file are either public or called by Mbed TLS, we can move all the declarations gated by `MBEDTLS_DECLARE_PRIVATE_IDENTIFIERS` to a private header file and adjust `#include` directives accordingly. This can be done by a script.
+
+### Analysis of privatization by header
+
+#### Headers that remain public
+
+The headers listed below declare functionality that has no PSA equivalent and that is desirable in 1.0/4.0. We will just move these headers to the public include directory. We may make further changes to some of these headers, but it is out of scope of this chapter.
+
+```
+asn1.h
+asn1write.h
+base64.h
+constant_time.h
+lms.h
+memory_buffer_alloc.h
+nist_kw.h
+pem.h
+pk.h
+platform.h
+platform_time.h
+platform_util.h
+psa_util.h
+threading.h
+```
+
+#### Headers that remain public for private macros
+
+The following headers solely define exposed macros, and must remain exposed. They can be
+
+```
+build_info.h
+config_adjust_legacy_from_psa.h
+config_adjust_psa_superset_legacy.h
+config_adjust_test_accelerators.h
+config_psa.h
+private_access.h
+```
+
+#### Privatization of built-in cryptographic headers with context types
```

Comment:
Minor: does not follow the same pattern "Headers that" as the sections before and after.